jobs.py

```diff
@@ -228,185 +204,157 @@ def get_next_job(self, lifo=False, device=None, worker=None):
-                            'from jobs where device=? order by istry desc, '
-                            'created %s' % order,
-                            (device,))]
-            if not jobs:
-                return None
-            next_job = jobs[0]
-            next_job['attempts'] += 1
-            next_job['last_attempt'] = datetime.datetime.now().isoformat()
-            c.execute('update jobs set attempts=?, last_attempt=? where id=?',
-                      (next_job['attempts'], next_job['last_attempt'],
-                       next_job['id']))
-            conn.commit()
-            next_job['tests'] = []
-            test_rows = [
-                {'name': test_row[0], 'guid': test_row[1]}
-                 for test_row in
-                 c.execute('select name, guid from tests where jobid=?',
-                           (next_job['id'],))]
-            for test_row in test_rows:
-                # Convert the list of test names in the job to a list
-                # of test objects and update their guid values.
-                for test in worker.tests:
-                    if test.name == test_row['name']:
-                        test.job_guid = test_row['guid']
-                        next_job['tests'].append(test)
-        except sqlite3.OperationalError:
-            logger.exception('jobs.get_next_job')
-            next_job = None
-        if next_job:
-            logger.debug('jobs.get_next_job: %s' % next_job)
-        return next_job
-
-    def cancel_job(self, test_name, test_guid, device=None):
-        logger.debug('jobs.cancel_job: %s %s %s' % (test_name, test_guid, device))
+        conn = self._conn()
+
+        # Find the ids of the jobs whose attempts exceed the maximum.
+        # First delete the associated tests, then the jobs.
+        job_cursor = self._execute_sql(
+            conn,
+            'select id from jobs where device=? and attempts>=?',
+            values=(device, self.MAX_ATTEMPTS))
+        job_ids = [job[0] for job in job_cursor]
+        for job_id in job_ids:
+            self._execute_sql(conn, 'delete from tests where jobid=?',
+                              values=(job_id,))
+        job_cursor.close()
+        self._execute_sql(
+            conn,
+            'delete from jobs where device=? and attempts>=?',
+            values=(device, self.MAX_ATTEMPTS))
+
+        self._commit_connection(conn)
+
+        job_cursor = self._execute_sql(
+            conn,
+            'select id,created,last_attempt,build_url,'
+            'build_id,changeset,tree,revision,revision_hash,'
+            'enable_unittests,attempts,instr(build_url,"try") as istry '
+            'from jobs where device=? order by istry desc, '
+            'created %s' % order,
+            values=(device,))
+
+        job_row = job_cursor.fetchone()
+        job_cursor.close()
+        if not job_row:
+            self._close_connection(conn)
```

Comment:
It feels like connection management should be simpler. In this function, and in other functions, there are a few calls to _close_connection(), due to multiple return paths. It would be nice if this could be re-arranged into something that looks like:
  conn = self._conn()
  <do everything you need to do>
  self._close_connection(conn)
Also, should we guard against exceptions? If something in here throws, will it take down autophone, or will we keep going, without closing the connection?
  conn = self._conn()
  try:
     _use_conn(conn)
  finally:
     self._close_connection(conn)
Maybe?